lib/blacklight/search_state.rb

```diff
@@ -4,10 +4,13 @@
 
 module Blacklight
   # This class encapsulates the search state as represented by the query
-  # parameters namely: :f, :q, :page, :per_page and, :sort
+  # parameters namely: :q, :f, :f_inclusive, :page, :per_page, :search_field and, :sort
   class SearchState
     attr_reader :blacklight_config, :params # Must be called blacklight_config, because Blacklight::Facet calls blacklight_config.
 
+    # These parameters represent the search state
+    PARAMETERS = %i[q f f_inclusive page per_page rows sort search_field controller format action clause op].freeze
```

Comment:
I think this will be  problematic for plugins that may rely on additional parameters?

e.g. https://github.com/geoblacklight/geoblacklight/blob/main/app/models/concerns/geoblacklight/bbox_filter_field.rb#L32, which creates and uses `params[:bbox]`.

Reply:
or range limit (`params[:range]`) - but it might be appropriate to suggest that BL8 include some type of accumulator pattern for SearchState to identify permitted params.

Reply:
it might also be possible to just ask the search state's current filters to represent themselves in a hash, starting with the query if present?

Reply:
For example,`SearchState::FilterField#add` does a lot of this work and might be usable  here.

Reply:
I moved this to the blacklight configuration and added documentation on how you might expand the  allowed list.

Reply:
@jcoyne can it automatically pick up anything added by the Blacklight 8 "new style" of SearchState plugin, as I think @barmintor is suggesting?  In addition to maybe allowing additional expansion of things not in SearchState, but it would probably be better if it automatically used SearchState?  I know this is a bit hard to figure out (I'm not sure how to do it myself), but I think making systems not consistent with each other just makes it more confusing. Maybe @barmintor could help?

Reply:
@jrochkind The SearchState::FilterField deals with selected facet values.  This PR doesn't touch the values (or anything at the second level), it just deals with which parameters are passed through at the top level.

Reply:
I actually took a swing at using the filter fields to do this - it's not crazy and I like it as a matter of responsibility - but it's expensive and there's a ton of tests that seem to expect search state to pass arbitrary fields around!

Reply:
(using `SearchState::FilterField#add` and the defined facet fields as the way to sanitize the non-query params, that is)

Reply:
@jcoyne So, sorry, I don't totally understand this either and might not be talking about it the right way. 

Fortunately, we have a fairly recent example written by someone who understood it to refer to to make this more concrete. 

Here is a PR to make `blacklight_range_limit` work with the "new" way of doing things that will continue working in BL8: https://github.com/projectblacklight/blacklight_range_limit/pull/179

blacklight_range_limit uses (since always) a new query param, `range` (which is a hash param with some nested hashes in it).  

That param really should be included  `search_state_fields ` and in SearchState#to_h -- it is part of the search state. (Sanity check: right?) This PR as it is would mean the blacklight_range_limit plugin would have to register that `range` param to have it properly treated. 

However, my understanding is that that PR is already registering `range` somewhere as a search_state field, in order to keep `blacklight_range_limit` working with BL8.  I cannot point to the line that's doing it because that PR is totally overwhelming me, and I don't understand it.   But I think it's already being registered somewhere... could that be re-used to form the basis for known "search state fields", so it don't need to be registered a second time?

I _think_ so. @cbeer or possibly @barmintor may be able to confirm or deny for sure? I am not aware of anyone else that understands this architecture -- I confess i do not.  But I'm thinking that it will make even harder to understand if this feature is not written to be consistent with it.  But I would love some feedback from @cbeer or @barmintor on that. 

`blacklight_range_limit` is just an example, I believe this will effect a host of other plugins and customizations. Including geoblacklight.  Some of which have already been refactored to use the patterns that `blacklight_range_limit` is using above. With this PR as it is, each of them would have to be returned to to also register their additional extension params with `config.search_state_fields` (how would a plugin even do that?).  However, if this logic here in this PR could be made to use the existing registrations that are being done already in current (eg) geoblacklight and blacklight_range_limit code, then the current code could Just Work.

I'm sorry this is confusing to talk about it... is this making any sense?  I think we could really use some help from @cbeer, or possibly @barmintor .

Reply:
worth noting that this is a BL8 change, so it's fine for existing plugins to have to document necessary configuration IMO.

Reply:
> worth noting that this is a BL8 change, so it's fine for existing plugins to have to document necessary configuration IMO.

@barmintor are you suggesting that rather than blacklight_range_limit automatically registering it's extension commands, it will document that a user installing it into a blacklight app should manually add a line to their CatalogController(s) to do `config.search_state_fields <<` with certain values? And if a future version of blacklight_range_limit changes what those values should be, it would just document in upgrade notes to go change that manually?

That seems unfortunate to me, and a departure from how plugins have worked. 

But I'll leave it up to you all who understand it better than I do(?) if you think that is the best that can be done or appropriate. 

I'm realizing my real distress is realizing that I completely don't understand how to write a blacklight plugin anymore that adds query params that effect Solr query.  And I think there are very very few people who do -- I'm not sure how many people could have written the blacklight_range_limit or geoblacklight PR's to update to BL8 -- possibly only one or two? 

And this PR adds an _additional_ customization point that any such plugin will have to touch, adding to that complexity.  When it's already (I think?) customizing the same semantic information elsewhere, so I thought it could be avoided.... but it may just be what it is, if you say so.  I am honestly concerned that I am not going to be able to update my local customizations because I will not be able to figure out how to, and there may be only one or two people who know how to. (@barmintor do you do contract work?)

Reply:
@jrochkind I'd like to point out that the existing way this works is broken for us.  It stores every field, which means it's useless. If you aren't happy with the solution I proposed, do you have an alternative fix to suggest?